tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/DynamoDb/BatchGetItemAsyncIntegration.cs

```diff
@@ -0,0 +1,87 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.ComponentModel;
+using System.Threading;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.DynamoDb
+{
+    /// <summary>
+    /// AWSSDK.DynamoDBv2 BatchGetItemAsync CallTarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "AWSSDK.DynamoDBv2",
+        TypeName = "Amazon.DynamoDBv2.AmazonDynamoDBClient",
+        MethodName = "BatchGetItemAsync",
+        ReturnTypeName = "System.Threading.Tasks.Task`1[Amazon.DynamoDBv2.Model.BatchGetItemResponse]",
+        ParameterTypeNames = new[] { "Amazon.DynamoDBv2.Model.BatchGetItemRequest", ClrNames.CancellationToken },
+        MinimumVersion = "3.0.0",
+        MaximumVersion = "3.*.*",
+        IntegrationName = AwsDynamoDbCommon.IntegrationName)]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class BatchGetItemAsyncIntegration
+    {
+        private const string Operation = "BatchGetItem";
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TBatchGetItemRequest">Type of the request object</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method</param>
+        /// <param name="request">The request for the DynamoDB operation</param>
+        /// <param name="cancellationToken">CancellationToken value</param>
+        /// <returns>CallTarget state value</returns>
+        internal static CallTargetState OnMethodBegin<TTarget, TBatchGetItemRequest>(TTarget instance, TBatchGetItemRequest request, CancellationToken cancellationToken)
+            where TBatchGetItemRequest : IBatchRequest, IDuckType
+        {
+            if (request.Instance is null)
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            var scope = AwsDynamoDbCommon.CreateScope(Tracer.Instance, Operation, out AwsDynamoDbTags tags);
+            if (request.RequestItems.Keys.Count != 1)
+            {
+                return new CallTargetState(scope);
+            }
+
+            // TableName tagging only when batch is from one table.
+            // Allocation is needed here due to RequestItems being
+            // Dictionary<string, Amazon.DynamoDBv2.Model.KeysAndAttributes>
+            var keys = request.RequestItems.Keys;
+            foreach (var key in keys)
+            {
+                var tableName = key as string;
+                tags.TableName = tableName;
+                var span = scope.Span;
+                span.ResourceName = $"{span.ResourceName} {tableName}";
+            }
```

Comment:
After some more trial and error, I think I found a solution that will work since we know these are both `System.Collections.Generic.Dictionary<string, SomethingElse>`. Can you see if this will still pass your local tests?

```suggestion
            // TableName tagging only when batch is from one table.
            var iterator = request.RequestItems.GetEnumerator();
            while (iterator.MoveNext())
            {
                var tableName = iterator.Key as string;
                tags.TableName = tableName;
                var span = scope.Span;
                span.ResourceName = $"{span.ResourceName} {tableName}";
            }
```

Reply:
I know we expect the key to always be a string, but I guess we should also do a null check on the `tableName` variable so we don't try to use a null value

Reply:
This works as expected! 

Amazing Zach, this is perfect to avoid the allocation! Much appreciated.

Do you think we should still do a null check on the key? Knowing that the dictionary already has at least one element, where it is guaranteed that string will be the element?

Reply:
Updated with null check and avoiding allocating into memory!